GVFS/GVFS.Common/InstallerPreRunChecker.cs

```diff
@@ -0,0 +1,243 @@
+using GVFS.Common;
+using GVFS.Common.Git;
+using GVFS.Common.Tracing;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+
+namespace GVFS.Upgrader
+{
+    public class InstallerPreRunChecker
+    {
+        private static readonly string[] BlockingProcessList = { "GVFS", "GVFS.Mount", "git", "ssh-agent", "bash", "gitk", "git-bash" };
+
+        private ITracer tracer;
+
+        public InstallerPreRunChecker(ITracer tracer)
+        {
+            this.tracer = tracer;
+        }
+
+        public bool TryRunPreUpgradeChecks(GitVersion gitVersion, out string error)
+        {
+            this.tracer.RelatedInfo("Checking if GVFS upgrade can be run on this machine.");
+
+            if (this.IsUnattended())
+            {
+                error = "`gvfs upgrade [--confirm]` is not supported in unattended mode";
+                this.tracer.RelatedError($"{nameof(TryRunPreUpgradeChecks)}: {error}");
+                return false;
+            }
+
+            if (this.IsDevelopmentVersion())
+            {
+                error = "Cannot run upgrade when development version of GVFS is installed.";
+                this.tracer.RelatedError($"{nameof(TryRunPreUpgradeChecks)}: {error}");
+                return false;
+            }
+
+            if (!this.IsGVFSUpgradeAllowed(out error))
+            {
+                this.tracer.RelatedError($"{nameof(TryRunPreUpgradeChecks)}: {error}");
+                return false;
+            }
+
+            this.tracer.RelatedInfo("Successfully finished pre upgrade checks. Okay to run GVFS upgrade.");
+
+            error = null;
+            return true;
+        }
+
+        public bool TryGetGitVersion(string gitPath, out GitVersion gitVersion, out string error)
+        {
+            error = null;
+            gitVersion = null;
+
+            string fakePath = string.Empty;
+            GVFSEnlistment enlistment = new GVFSEnlistment(fakePath, fakePath, gitPath, fakePath);
+            GitProcess.Result versionResult = GitProcess.Version(enlistment);
+            string version = versionResult.Output;
+
+            if (!GitVersion.TryParseGitVersionCommandResult(version, out gitVersion))
+            {
+                error = "Unable to determine installed git version. " + version;
+                return false;
+            }
+
+            return true;
+        }
+
+        public virtual bool TryGetGitVersion(out GitVersion gitVersion, out string error)
+        {
+            return this.TryGetGitVersion(
+                GVFSPlatform.Instance.GitInstallation.GetInstalledGitBinPath(), 
+                out gitVersion, 
+                out error);
+        }
+
+        public bool TryMountAllGVFSRepos(out string error)
+        {
+            return this.TryRunGVFSWithArgs("service --mount-all", out error);

```

Comment:
As discussed this will cause a cyclic dependency.

Alternate option - This method does Mount/Re-Mount, it is not related directly to Installer precheck. Mount related methods can be moved out of this class into GVFS.Upgrader project. Add the dependencies required to call GVFSVerb.Execute inside GVFS.Upgrader and replace the CLI invocation with in proc. I can do this in a future PR.